photosynthesis_metrics/gmsd.py

```diff
@@ -0,0 +1,253 @@
+    trgt_grad_y = F.conv2d(target, kernel_y)
+    
+    # Compute grad direction
+    pred_grad = torch.sqrt(pred_grad_x ** 2 + pred_grad_y ** 2)
+    trgt_grad = torch.sqrt(trgt_grad_x ** 2 + trgt_grad_y ** 2)
+    
+    # Compute GMS
+    gms = (2.0 * pred_grad * trgt_grad + EPS) / (pred_grad ** 2 + trgt_grad ** 2 + EPS)
+    mean_gms = torch.mean(gms, dim=[1, 2, 3], keepdims=True)
+
+    # Compute GMSD
+    gmsd = torch.pow(gms - mean_gms, 2).mean(dim=[1, 2, 3], keepdims=True).sqrt()
+    
+    if reduction == 'mean':
+        return gmsd.mean()
+    elif reduction == 'sum':
+        return gmsd.sum()
+    else:
+        return gmsd
+
+
+class GMSDLoss(_Loss):
+    r"""Creates a criterion that measures Gradient Magnitude Similarity Deviation
+    between each element in the input and target.
+
+    Args:
+        reduction: Specifies the reduction to apply to the output:
+            ``'none'`` | ``'mean'`` | ``'sum'``. ``'none'``: no reduction will be applied,
+            ``'mean'``: the sum of the output will be divided by the number of
+            elements in the output, ``'sum'``: the output will be summed. Default: ``'mean'``
+        data_range: The difference between the maximum and minimum of the pixel value,
+            i.e., if for image x it holds min(x) = 0 and max(x) = 1, then data_range = 1.
+            The pixel value interval of both input and output should remain the same.
+            
+    Reference:
+        Wufeng Xue et al. Gradient Magnitude Similarity Deviation (2013)
+        https://arxiv.org/pdf/1308.3052.pdf
+        
+    """
+
+    def __init__(self, reduction: str = 'mean', data_range: Union[int, float] = 1.) -> None:
+        super().__init__()
+
+        # Generic loss parameters.
+        self.reduction = reduction
+
+        # Loss-specific parameters.
+        self.data_range = data_range
+
+    def forward(self, prediction: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+        r"""Computation of Gradient Magnitude Similarity Deviation (GMSD) as a loss function.
+
+        Args:
+            prediction: Tensor of prediction of the network.
+            target: Reference tensor.
+
+        Returns:
+            Value of GMSD loss to be minimized. 0 <= GMSD loss <= 1.
+        """
+        _validate_input(x=prediction, y=target, scale_weights=None)
+        prediction, target = _adjust_dimensions(x=prediction, y=target)
+
+        return self.compute_metric(prediction, target)
+    
+    def compute_metric(self, prediction: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+        
+        if self.data_range == 255:
+            prediction = prediction / 255.
+            target = target / 255.
+
+        # Average by 2x2 filter and downsample
+        padding = (prediction.shape[2] % 2, prediction.shape[3] % 2)
+        prediction = F.avg_pool2d(prediction, kernel_size=2, stride=2, padding=padding)
+        target = F.avg_pool2d(target, kernel_size=2, stride=2, padding=padding)
```

Comment:
According to [last paper](http://www.cse.ust.hk/~psander/docs/gradsim.pdf) there is no need in downsample for GMSD. 
Also, it is stated that first GMDS at multiscale approach is calculated for original sized images. In current implementation, we do downsampling first, then calculate GMDS.

Reply:
See original paper for [GMSD]( https://arxiv.org/pdf/1308.3052.pdf) Section 3.B 
>In addition, as in the implementations of SSIM [8] and FSIM [7], the images r and d are first filtered by a 2×2 average filter, and then down-sampled by a factor of 2

They also do it in their MATLAB code. Another reason for this is weights for MS-GMSD, where the biggest weight in on scale 2, which is exactly blurred and downsampled image.